CRM/Member/Form/Membership.php

```diff
@@ -1723,11 +1723,87 @@ public function submit() {
     }
 
     $isRecur = CRM_Utils_Array::value('is_recur', $params);
+    $this->updateContributionOnMembershipTypeChange($params, $membership);
     $this->setStatusMessage($membership, $endDate, $receiptSent, $membershipTypes, $createdMemberships, $isRecur, $calcDates, $mailSend);
     return $createdMemberships;
   }
 
   /**
+   * Update related contribution of a membership if update_contribution_on_membership_type_change
+   *   contribution setting is enabled and type is changed on edit
+   *
+   * @param array $inputParams
+   *      submitted form values
+   * @param CRM_Member_DAO_Membership $membership
+   *     Updated membership object
+   *
+   */
+  protected function updateContributionOnMembershipTypeChange($inputParams, $membership) {
+    if (Civi::settings()->get('update_contribution_on_membership_type_change') &&
+      ($this->_action & CRM_Core_Action::UPDATE) && // on update
+      $this->_id && // if ID is present
+      !in_array($this->_memType, $this->_memTypeSelected) && // if selected membership doesn't match with earlier membership
+      !CRM_Utils_Array::value('is_recur', $inputParams) // don't support recurring contribution
```

Comment:
What is behaviour if these conditions are not met, eg the membership is paid by recurring? We should fall back on existing warnings rather than display new info message? To do that don't we need to return a success or failure arg or something to allow that to happen?